analytics/certificates-jan2025.md

```diff
@@ -0,0 +1,115 @@
+    - Message: the 32-byte hash of the Praos block being boosted is also included in the certificate.
+- Identity of voters
+    - Persistent voters are encoded in a bitset of size $m$, occupying $\left\lceil m / 8 \right\rceil$ bytes.
+    - Non-persistent voters are encoded by their Pool ID (or equivalent), occupying 32 bytes each and hence $32 \cdot (n - m)$ bytes total.
+    - Alternatively, all possible voters could be assigned bits in a bitset, with size $\left\lceil N_\text{pools} / 8 \right\rceil$.
+- Eligibility proof
+    - Persistent voters are eligible by definition (by virtue of their stake in the epoch), so no proof is needed.
+    - Non-persistent voters prove eligibility with a 96-byte BLS signature on the message, occupying $96 \cdot (n - m)$ bytes total.
+- Aggregate signatures
+    - Signed message: This aggregate BLS signature on the message is 96 bytes.
+    - Signed election proofs: Perhaps not strictly necessary, but another 96 byte BLS signature can attest to the proof of the eligibility, see **BLS.BSig** in [the Leios paper](https://iohk.io/en/research/library/papers/high-throughput-blockchain-consensus-under-realistic-network-assumptions/).
+    
+Thus the total certificate size is
+
+$$
+\text{certificate bytes} = 256 + \left\lceil \frac{m}{8} \right\rceil + 128 \cdot (n - m)
+$$
+
+but not including any minor overhead arising from CBOR serialization. As noted previously, only a quorum of votes actually needs to be recorded.
+
+
+### Benchmarks in Rust
+
+- Sortition
+    - Persistent voters: not benchmarked because it would be reasonably fast and only computed once per epoch
+    - Non-persistent voters: 4.1 µs
+- Vote
+    - Verify the proof of key possession: 1.5 ms/key
+    - Generate vote: 280 µs/vote
+    - Verify vote: 1.4 ms/vote
+- Certificate
+    - Generate certificate for a 750-vote quorum: 125 ms/cert
+    - Verify certificate for a 750-vote quorum: 225 ms/cert
+
+
+### Realistic stake distribution
+
+The following plots show number of persistent votes and votes, along with certificate size, for the `mainnet` stake distribution of Epoch 535. The dashed line in the first two plots has slope one, so the gap between it and the solid line indicates the number of non-persistent voters or votes. The certificate-size plot does not take into account a potential reduction in certificate size from omitting votes in excess of a quorum.
+
+| Voters                                            | Votes                                           | Certificate size                                          |
+| ------------------------------------------------- | ----------------------------------------------- | --------------------------------------------------------- |
+| ![Fiat-accompli voters](fiat-accompli-voters.svg) | ![Fiat-accompli votes](fiat-accompli-votes.svg) | ![Fiat-accompli certificate size](fiat-accompli-cert.svg) |
```

Comment:
I assume that you plot expected numbers (of voters/votes/cert size) against the target committee size (i.e. the $n$ parameter in wFA) here.

I would assume that the vote graph is identical to the $x=y$ axis, as the committee tuple $C$ returned by $\mathrm{wFA}^{\mathrm{LS}}$ has expected size ${|C|} = n$: First, $\mathrm{wFA}$ chooses $n_1<n$ committee seats deterministically, and then invokes $\mathrm{LS}$ with target committee size $n-n_1$. $\mathrm{LS}$ will then return $n-n_1$ committee members on average (potentially with duplicates), so on average, we allocated $n_1 + (n-n_1) = n$ committee seats.

Am I missing something?

Reply:
Sorry, the captions are a bit sloppy. Here are how the *y*-axes are calculated:

- `Persistent voters` = $`i^*`$
- `Persistent votes` = $`n \cdot (1 - \rho^*)`$, so this really should be called the vote-equivalent weight of the persistent voters

I don't think the "votes" plot adds value, so I'll delete it.

Reply:
Deleted in 47b6c9944c5299b137b56051323dfe0da28fa330.